pandas/core/dtypes/cast.py

```diff
@@ -275,12 +276,15 @@ def maybe_cast_result(
     dtype = maybe_cast_result_dtype(dtype, how)
 
     if not is_scalar(result):
-        if is_extension_array_dtype(dtype) and dtype.kind != "M":
-            # The result may be of any type, cast back to original
-            # type if it's compatible.
-            if len(result) and isinstance(result[0], dtype.type):
-                cls = dtype.construct_array_type()
-                result = maybe_cast_to_extension_array(cls, result, dtype=dtype)
+        if (
```

Comment:
again instead of expanding this check, i would completely remove it; it be encompassed in maybe_cast_result_type, which is the point of that function

Reply:
I'm not sure if we can completely move this logic into `maybe_cast_result_dtype`; e.g., for something like `agg(pd.Series.nunique)` performed on a categorical, we end up with integer counts with an original dtype of categorical, but we don't actually know "how" we got there, so we can't say that the dtype should still be integer based on the kind of operation that was performed.

The datetime check seems to be another issue with not being able to introspect a user-defined function like `agg(lambda g: g.iloc[0].year)` and know that the output should still be an int and not a datetime.

Reply:
this is tagged as 1.0.4 at the moment just in case we do another patch release. If so, the changes in this PR should be kept to a minimum?